x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/InternalCategorizationAggregation.java

```diff
@@ -0,0 +1,428 @@
+        int maxDepth,
+        double similarityThreshold,
+        Map<String, Object> metadata,
+        List<Bucket> buckets
+    ) {
+        super(name, metadata);
+        this.buckets = buckets;
+        this.maxChildren = maxChildren;
+        this.maxDepth = maxDepth;
+        this.similarityThreshold = similarityThreshold;
+        this.minDocCount = minDocCount;
+        this.requiredSize = requiredSize;
+    }
+
+    public InternalCategorizationAggregation(StreamInput in) throws IOException {
+        super(in);
+        this.maxChildren = in.readVInt();
+        this.maxDepth = in.readVInt();
+        this.similarityThreshold = in.readDouble();
+        this.buckets = in.readList(Bucket::new);
+        this.requiredSize = readSize(in);
+        this.minDocCount = in.readVLong();
+    }
+
+    @Override
+    public InternalCategorizationAggregation create(List<Bucket> buckets) {
+        return new InternalCategorizationAggregation(
+            name,
+            requiredSize,
+            minDocCount,
+            maxChildren,
+            maxDepth,
+            similarityThreshold,
+            super.metadata,
+            buckets
+        );
+    }
+
+    @Override
+    public Bucket createBucket(InternalAggregations aggregations, Bucket prototype) {
+        return new Bucket(prototype.key, prototype.docCount, aggregations);
+    }
+
+    @Override
+    protected Bucket reduceBucket(List<Bucket> buckets, ReduceContext context) {
+        throw new IllegalArgumentException("For optimization purposes, typical bucket path is not supported");
+    }
+
+    // This is for buckets that share the same key
+    protected DelayedCategorizationBucket delayReduction(List<Bucket> buckets) {
+        assert buckets.size() > 0;
+        long docCount = 0;
+        for (Bucket bucket : buckets) {
+            docCount += bucket.docCount;
+        }
+        return new DelayedCategorizationBucket(buckets.get(0).key, buckets, docCount);
+    }
+
+    @Override
+    public List<Bucket> getBuckets() {
+        return buckets;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return CategorizeTextAggregationBuilder.NAME;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeVInt(maxChildren);
+        out.writeVInt(maxDepth);
+        out.writeDouble(similarityThreshold);
+        out.writeList(buckets);
+        writeSize(requiredSize, out);
+        out.writeVLong(minDocCount);
+    }
+
+    @Override
+    public InternalAggregation reduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        CategorizationTokenTree categorizationTokenTree = new CategorizationTokenTree(maxChildren, maxDepth, similarityThreshold);
+        // TODO: Could we do a merge sort similar to terms?
+        //  It would require us returning partial reductions sorted by key, not by doc_count
+        // First, make sure we have all the counts for equal log groups
```

Comment:
The `terms` agg does that! Check out `TopBucketBuilder` and its callers. It's kind of a wild ride.

Reply:
@nik9000 yes, I attempted it that was at first, but it causes some conflict of interests in partial merging. 

Terms will still return buckets sorted by key in a partial reduce scenario, but I am not sure we want that for this aggregation. It seems to me that we would only want the top buckets (sorted by doc count). 

I might be missing something here.